File: bindings/python/src/zlink/contracts/service/spot/spot_operations.py
# SPDX-License-Identifier: MPL-2.0

from typing import Protocol, runtime_checkable


@runtime_checkable
class _FluentMessageOp(Protocol):
    """Shared fluent builder for operations that carry message parts.

    Submitting consumes the added parts: on a successful submit each part's
    payload is moved into the transport and the managed message is left empty,
    so a part must not be reused after a successful submit.
    """

    def message(self, payload):
        """Add a message part. It is consumed on a successful submit (see the
        class docstring)."""
        ...

    def messages(self, *payloads):
        """Add several message parts in order; each is consumed on a successful
        submit."""
        ...

    def flags(self, flags):
        """Set the send flags applied at submit time."""
        ...


@runtime_checkable
class SendOp(_FluentMessageOp, Protocol):
    """A multipart send builder: add parts, then submit."""

    def submit(self):
        """Submit the accumulated parts. Return ``True`` when queued, and
        ``False`` only when ``DONT_WAIT`` is set and the send would have blocked
        (back-pressure)."""
        ...


@runtime_checkable
class RequestOp(_FluentMessageOp, Protocol):
    """A request builder: add parts, then submit and await a reply."""

    def timeout(self, timeout):
        """Set how long the request waits for a reply before timing out."""
        ...

    def submit_async(self):
        """Submit the request and asynchronously return the reply parts, which
        the caller owns."""
        ...

    def submit(self, callback):
        """Submit the request; the reply is delivered later to ``callback``,
        which owns the reply parts."""
        ...


@runtime_checkable
class RequestCallbackOp(_FluentMessageOp, Protocol):
    """Callback-submission stage of a request (reached after setting flags)."""

    def timeout(self, timeout):
        """Set how long the request waits for a reply before timing out."""
        ...

    def submit(self, callback):
        """Submit the request; the reply is delivered later to ``callback``."""
        ...


@runtime_checkable
class ReplyOp(_FluentMessageOp, Protocol):
    """A reply builder: add parts, then submit."""

    def submit(self):
        """Submit the accumulated reply parts."""
        ...


@runtime_checkable
class ActorJoinOp(_FluentMessageOp, Protocol):
    """An actor-join builder: add parts, then submit and await the result."""

    def timeout(self, timeout):
        """Set how long the operation waits for completion before timing out."""
        ...

    def submit_async(self):
        """Submit and asynchronously return the join result and reply parts,
        which the caller owns."""
        ...

    def submit(self, callback):
        """Submit; the result and reply parts are delivered later to
        ``callback``."""
        ...


@runtime_checkable
class ActorJoinCallbackOp(_FluentMessageOp, Protocol):
    """Callback-submission stage of an actor join."""

    def timeout(self, timeout):
        """Set how long the operation waits for completion before timing out."""
        ...

    def submit(self, callback):
        """Submit; the result is delivered later to ``callback``."""
        ...


@runtime_checkable
class ActorJoinEntrySpotOp(Protocol):
    """A builder for joining an actor through an entry spot."""

    def timeout(self, timeout):
        """Set how long the operation waits for completion before timing out."""
        ...

    def submit_async(self):
        """Submit and asynchronously return the result."""
        ...

    def submit(self, callback):
        """Submit; the result is delivered later to ``callback``."""
        ...


@runtime_checkable
class ActorJoinReplyOp(_FluentMessageOp, Protocol):
    """A builder for replying to an actor-join request; a zero-part reply is
    allowed."""

    def submit(self):
        """Submit the actor-join reply."""
        ...


@runtime_checkable
class ActorLeaveOp(Protocol):
    """A payload-less actor operation builder (leave, destroy, bind, unbind)."""

    def timeout(self, timeout):
        """Set how long the operation waits for completion before timing out."""
        ...

    def submit_async(self):
        """Submit and asynchronously return the reply parts, which the caller
        owns."""
        ...

    def submit(self, callback):
        """Submit; the reply is delivered later to ``callback``."""
        ...


@runtime_checkable
class ActorDestroyOp(ActorLeaveOp, Protocol):
    """A builder for destroying an actor."""

    ...


@runtime_checkable
class ActorLookupOp(Protocol):
    """A builder for looking up a remote actor's reference."""

    def timeout(self, timeout):
        """Set how long the lookup waits for completion before timing out."""
        ...

    def submit_async(self):
        """Submit and asynchronously return the lookup result."""
        ...

    def submit(self, callback):
        """Submit; the result is delivered later to ``callback``."""
        ...


@runtime_checkable
class ActorBindOp(ActorLeaveOp, Protocol):
    """A builder for binding an actor to a session."""

    ...


@runtime_checkable
class ActorUnbindOp(ActorLeaveOp, Protocol):
    """A builder for removing an actor's session binding."""

    ...


__all__ = [
    "ActorBindOp",
    "ActorDestroyOp",
    "ActorJoinCallbackOp",
    "ActorJoinEntrySpotOp",
    "ActorJoinOp",
    "ActorJoinReplyOp",
    "ActorLeaveOp",
    "ActorLookupOp",
    "ActorUnbindOp",
    "ReplyOp",
    "RequestCallbackOp",
    "RequestOp",
    "SendOp",
]

File: bindings/node/src/zlink/contracts/service/spot/spot_operations.ts
// SPDX-License-Identifier: MPL-2.0

import type { Message } from '../../messaging';
import type { SendFlags } from '../../sockets/socket_constants';
import type { Flaggable, PartBuilder, ReplyHandler, Timeoutable } from '../../messaging/operations';
export type * from '../../messaging/operations';
import type {
  ActorJoinEntrySpotResult,
  ActorJoinResult,
  ActorLookupResult,
} from './spot_models';

/** Invoked with the result of an actor join and its reply parts, which the callback owns. */
export type ActorJoinHandler = (result: ActorJoinResult, parts: Message[]) => void;
/** Invoked with the result of an actor entry-spot join and its reply parts, which the callback owns. */
export type ActorJoinEntrySpotHandler = (result: ActorJoinEntrySpotResult, parts: Message[]) => void;
/** Invoked with the result of an actor lookup. */
export type ActorLookupHandler = (result: ActorLookupResult) => void;

/** An actor-join builder: add parts, then submit and await the result. Parts are consumed on a successful submit. */
export interface ActorJoinOperation extends PartBuilder<ActorJoinSubmitOperation> {}

/** Accepts further parts, timeout, flags, and the terminal submit of an actor join. */
export interface ActorJoinSubmitOperation
  extends PartBuilder<ActorJoinSubmitOperation>, Timeoutable<ActorJoinSubmitOperation> {
  /** Set the send flags and narrow the builder to callback submission. */
  flags(flags: SendFlags): ActorJoinCallbackSubmitOperation;
  /** Submit and return the result and reply parts, which the caller owns. */
  submit(): Promise<{ result: ActorJoinResult; parts: Message[] }>;
  /** Submit; the result and reply parts are delivered to `callback`. */
  submit(callback: ActorJoinHandler): boolean;
}

/** Callback-submission stage of an actor join. */
export interface ActorJoinCallbackSubmitOperation
  extends PartBuilder<ActorJoinCallbackSubmitOperation>,
    Timeoutable<ActorJoinCallbackSubmitOperation>,
    Flaggable<ActorJoinCallbackSubmitOperation> {
  /** Submit; the result and reply parts are delivered to `callback`. */
  submit(callback: ActorJoinHandler): boolean;
}

/** A builder for joining an actor through an entry spot. */
export interface ActorJoinEntrySpotOperation
  extends PartBuilder<ActorJoinEntrySpotOperation>,
    Timeoutable<ActorJoinEntrySpotOperation>,
    Flaggable<ActorJoinEntrySpotOperation> {
  /** Submit and return the result and reply parts, which the caller owns. */
  submit(): Promise<{ result: ActorJoinEntrySpotResult; parts: Message[] }>;
  /** Submit; the result and reply parts are delivered to `callback`. */
  submit(callback: ActorJoinEntrySpotHandler): boolean;
}

/** A builder for replying to an actor-join request; a zero-part reply is allowed. */
export interface ActorJoinReplyOperation extends PartBuilder<ActorJoinReplyOperation> {
  /** Submit the actor-join reply. */
  submit(): void;
}

/** A builder for an actor leave. */
export interface ActorLeaveOperation extends Timeoutable<ActorLeaveOperation> {
  /** Submit and return the reply parts, which the caller owns. */
  submit(): Promise<Message[]>;
  /** Submit; the reply is delivered to `callback`. */
  submit(callback: ReplyHandler): boolean;
}

/** A builder for destroying an actor. */
export interface ActorDestroyOperation extends Timeoutable<ActorDestroyOperation> {
  /** Submit and return the reply parts, which the caller owns. */
  submit(): Promise<Message[]>;
  /** Submit; the reply is delivered to `callback`. */
  submit(callback: ReplyHandler): boolean;
}

/** A builder for looking up a remote actor's reference. */
export interface ActorLookupOperation extends Timeoutable<ActorLookupOperation> {
  /** Submit and return the lookup result. */
  submit(): Promise<ActorLookupResult>;
  /** Submit; the result is delivered to `callback`. */
  submit(callback: ActorLookupHandler): boolean;
}

/** A builder for binding an actor to a session. */
export interface ActorBindOperation extends Timeoutable<ActorBindOperation> {
  /** Submit and return the reply parts, which the caller owns. */
  submit(): Promise<Message[]>;
  /** Submit; the reply is delivered to `callback`. */
  submit(callback: ReplyHandler): boolean;
}

/** A builder for removing an actor's session binding. */
export interface ActorUnbindOperation extends Timeoutable<ActorUnbindOperation> {
  /** Submit and return the reply parts, which the caller owns. */
  submit(): Promise<Message[]>;
  /** Submit; the reply is delivered to `callback`. */
  submit(callback: ReplyHandler): boolean;
}
